src/Microsoft.VisualStudio.Threading/JoinableTaskFactory.cs

```diff
@@ -905,6 +895,12 @@ public void GetResult()
                 // changes they apply (including SynchronizationContext) when they complete, thanks to the way .NET 4.5 works.
                 var syncContext = this.job != null ? this.job.ApplicableJobSyncContext : this.jobFactory.ApplicableJobSyncContext;
                 syncContext.Apply();
+
+                // Cancel if requested, even if we arrived on the main thread.
+                // Unlike most async methods where throwing OperationCanceledException after completing the work may not be a good idea,
+                // SwitchToMainThreadAsync is a scheduler method, and always precedes some work by the caller that almost certainly should
+                // not be carried out if cancellation was requested.
+                this.cancellationToken.ThrowIfCancellationRequested();
```

Comment:
📝 The reordering here can actually produce an observable side effect. Awaiting the pre-canceled token on the UI thread will still apply the synchronization context from the job, which could differ from the synchronization context prior to the caller. The most likely case where code could be impacted is something like this:

```csharp
try {
  await jtf.SwitchToMainThreadAsync(cancellationToken);
} catch (OperationCanceledException) {
}

// This call implicitly uses ConfigureAwait(true), which would continue on a background thread
// after yielding prior to this change, but now switches to the main thread even on the cancellation path.
await Method();
```

Reply:
Ooh! nice catch. The reordering was intentional so that we don't throw cancellation on the main thread without having set the sync context first. But I inadvertently applied it to a background thread too. I'll fix that!

Reply:
Ah, no actually I think it's correct as-is. We don't set the main thread SynchronizationContext on a background thread. We're applying `ApplicableJobSyncContext` which automatically detects and applies the right sync context for the thread. On a background thread this isn't necessarily null (e.g. you're in a JTF.Run delegate that started on a background thread, so a sync context may apply even on the background thread).

I'm adding tests for this anyway though.